### ISSUE TITLE
Can't run new containers - grpc: the connection is unavailable.

### ISSUE THREAD

**Issue by ebeigarts** [docker/docker#31849]
**Description**

I have to restart docker every day or so, to be able to start new containers.
I think these errors started happening recently after upgrading docker.

```
$ docker run -i -t busybox bash
docker: Error response from daemon: grpc: the connection is unavailable.
```

**Steps to reproduce the issue:**

1. Wait a random time?
2. Run container and get `grpc: the connection is unavailable.`
3. Restart docker `systemctl restart docker`
5. Run container successfully

**Output of `docker version`:**

```
Client:
 Version:      17.03.0-ce
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 10:53:29 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.03.0-ce
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 10:53:29 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 42
 Running: 4
 Paused: 0
 Stopped: 38
Images: 13
Server Version: 17.03.0-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 266
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: N/A (expected: 977c511eda0925a723debdc94d09459af49d082a)
runc version: a01dafd48bc1c7cc12bdb01206f9fea7dd6feb70
init version: 949e6fa
Kernel Version: 3.16.0-4-amd64
Operating System: Debian GNU/Linux 8 (jessie)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.826 GiB
Name: intranet
ID: VW6E:AS3L:73N3:WXH7:3DNF:YXOQ:G7AP:T2YD:ZXMJ:JY5F:24NG:3KSI
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Http Proxy: http://user:pass@10.101.0.20:3128/
Username: ...
Registry: https://index.docker.io/v1/
WARNING: No memory limit support
WARNING: No swap limit support
WARNING: No kernel memory limit support
WARNING: No oom kill disable support
WARNING: No cpu cfs quota support
WARNING: No cpu cfs period support
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

OS: Debian GNU/Linux 8

**Comment by mlaventure** [comment]
@ebeigarts please provide the content of the docker logs around the failure it should hopefully tell us what happenned

**Comment by ebeigarts** [comment]
`journalctl -u docker.service`: https://gist.github.com/ebeigarts/8ee303a65dd8e03c88f0328d2ab378a1

This is the time when I restarted the docker:

```
Mar 15 15:38:16 intranet systemd[1]: Stopping Docker Application Container Engine...
Mar 15 15:38:32 intranet systemd[1]: Starting Docker Application Container Engine...
```

**Comment by mlaventure** [comment]
@ebeigarts I do not see the grpc error in the log you've provided.

If you're still getting the error can you activate debug mode in the daemon, try again and provide the new log?

You can activate debug in a running daemon by following these instructions: 
1. Update `/etc/docker/daemon.json` to add `"debug": true`. E.g.:
```json
{
  "debug": true
}
```
2. Reload the daemon configuration: `kill -HUP $(cat /run/docker.pid)`

**Comment by ebeigarts** [comment]
Thanks, if this will happen again, I will turn on the debugging and wait for it to happen again. You can probably close this for now, I will reopen the issue If this will happen again

**Comment by mlaventure** [comment]
Ok, closing then. 

Thanks!

**Comment by dikderoy** [comment]
@mlaventure 
I've catched same error, performed "enabling debug" ops you described. but no additional info showed up...:

_with debug mode enabled on server and client_

```
# docker --debug run --rm wk-devops/cb-env-pipeline-base:latest bash
DEBU[0000] [hijack] End of stdout
docker: Error response from daemon: grpc: the connection is unavailable.
```

in service logs I see some suspicious lines (but dated a week ago)

`journalctl -u docker.service`

```
May 18 04:52:27 ip-172-31-55-2.ec2.internal dockerd[3821]: time="2017-05-18T04:52:27.336082305-04:00" level=info msg="Layer sha256:2d68ee885292be0a33d15cae643068d7e7ccab6b691cacc153737e718938eea7 cleaned
May 18 04:53:07 ip-172-31-55-2.ec2.internal dockerd[3821]: time="2017-05-18T04:53:07.034879973-04:00" level=info msg="Layer sha256:f5f659fd3851889516a49524e8ad3713ac6c087ea5ac4b47325f05e2b16e3183 cleaned
May 24 06:28:35 ip-172-31-55-2.ec2.internal dockerd[3821]: 2017/05/24 06:27:40 grpc: Server.processUnaryRPC failed to write status stream error: code = 4 desc = "context deadline exceeded"
May 24 06:31:59 ip-172-31-55-2.ec2.internal dockerd[3821]: time="2017-05-24T06:29:49.232755070-04:00" level=error msg="libcontainerd: failed to receive event from containerd: rpc error: code = 13 desc = t
~
```

`docker version`

```
Client:
 Version:      17.04.0-ce
 API version:  1.28
 Go version:   go1.7.5
 Git commit:   4845c56
 Built:        Mon Apr  3 18:01:50 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.04.0-ce
 API version:  1.28 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   4845c56
 Built:        Mon Apr  3 18:01:50 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

`docker info`

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 90
Server Version: 17.04.0-ce
Storage Driver: overlay
 Backing Filesystem: xfs
 Supports d_type: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary:
containerd version: N/A (expected: 422e31ce907fd9c3833a38d7b8fdd023e5a76e73)
runc version: 9c2d8d184e5da67c95d601382adf14862e4f2228
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-514.6.1.el7.x86_64
Operating System: Red Hat Enterprise Linux Server 7.3 (Maipo)
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 1.795GiB
Name: ip-172-31-55-2.ec2.internal
ID: RVXZ:GRNV:SMDF:SUJJ:4ARY:7GUP:VU2O:2IZ3:LADR:YCRW:LCQZ:BSRW
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 14
 Goroutines: 18
 System Time: 2017-06-01T00:57:39.579005502-04:00
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional info**
AWS EC2

```
# uname -a
Linux 3.10.0-514.6.1.el7.x86_64 #1 SMP Sat Dec 10 11:15:38 EST 2016 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by mlaventure** [comment]
@dikderoy may need more of the log to see what's happening.

It looks like `docker-containerd` is not running on your machine,

**Comment by dikderoy** [comment]
here is the entire log. I must mention - it was a Jenkins machine - mostly used for builds, not running containers on a daily basis. and builds was working as I remember (at least between 24th and 31th of may for sure)
[docker.txt](https://github.com/moby/moby/files/1046102/docker.txt)

**Comment by ccactus** [comment]
I'm still seeing this on 17.03.1-ce, build c6d412e
```bash
docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 77
Server Version: 17.03.1-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 72
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 4ab9917febca54791c5f071a9d1f404867857fcc
runc version: 54296cf40ad8143b62dbcaa1d90e520a2136ddfe
init version: 949e6fa
Kernel Version: 3.16.0-4-amd64
Operating System: Debian GNU/Linux 8 (jessie)
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 15.71 GiB
Name: client-docker-3
ID: TRO4:GAYI:HONM:DIGJ:GG5J:XTIY:FSQF:X6LH:7ZJY:UEZD:DG3X:MPRL
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No memory limit support
WARNING: No swap limit support
WARNING: No kernel memory limit support
WARNING: No oom kill disable support
WARNING: No cpu cfs quota support
WARNING: No cpu cfs period support
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

It happens intermittently on our build machines. At first it looked like a re-install (and change from using docker-engine to docker-ce) fixed it longer term but one of the boxes is still seeing it. I'm going to enable debug but haven't got any logs for that yet. A couple of things I've noticed that often occur at the same time:

- Whenever it has happened there has been at least one container in "Created" state but that can't be removed but returns the error message `Error response from daemon: Driver aufs failed to remove root filesystem 3fc4735cf8b6d523544c7995d28230061a003f143cd9a4546f5362f18c7ab620: no such file or directory`. The latest one was at build stage interrupted by a "transport is closing" message, unfortunately the log I have doesn't say more than that. 

- The output from systemctl at the time of grpc error said:
```bash
 sudo systemctl status docker
● docker.service - Docker Application Container Engine
Loaded: loaded (/lib/systemd/system/docker.service; enabled
   Active: active (running) since Thu 2017-06-22 11:31:52 BST; 23h ago
     Docs: https://docs.docker.com
 Main PID: 10864 (dockerd)
   CGroup: /system.slice/docker.service
           ├─10864 /usr/bin/dockerd -H fd://
           └─14658 docker-containerd-shim
3fc4735cf8b6d523544c7995d28230061a003f143cd9a4546f5362f18c7ab620 \n/var/run/docker/libcontainerd/3fc4735cf8b6d523544c7995d28230061a003f143cd9a4546f5362f18c7ab620 docker-runc 

Jun 22 18:21:52 client-docker-3 dockerd[10864]: time="2017-06-22T18:21:52.251112724+01:00" level=error msg="Handler for POST /v1.27/containers/create returned error: No such image: acano/test:latest"
Jun 22 18:21:53 client-docker-3 dockerd[10864]: time="2017-06-22T18:21:53.604758851+01:00" level=error msg="Not continuing with pull after error: errors:\ndenied: requested access to the resource is denied\n...ion required\n"
Jun 22 18:21:53 client-docker-3 dockerd[10864]: time="2017-06-22T18:21:53.604803444+01:00" level=info msg="Ignoring extra error returned from registry: unauthorized: authentication required"
Jun 22 18:21:53 client-docker-3 dockerd[10864]: time="2017-06-22T18:21:53.604853666+01:00" level=info msg="Translating \"denied: requested access to the resource is denied\" to \"repository acano/test not fo... pull access\""
Jun 22 18:21:53 client-docker-3 dockerd[10864]: time="2017-06-22T18:21:53.604886184+01:00" level=error msg="Handler for POST /v1.27/images/create returned error: repository acano/test not found: does not exi...no pull access"
Jun 22 21:59:45 client-docker-3 dockerd[10864]: time="2017-06-22T21:59:45.912250087+01:00" level=error msg="containerd: notify OOM events" error="cgroup path for memory not found"
Jun 23 10:59:36 client-docker-3 systemd[1]: [/lib/systemd/system/docker.service:24] Unknown lvalue 'Delegate' in section 'Service'
Jun 23 10:59:43 client-docker-3 systemd[1]: [/lib/systemd/system/docker.service:24] Unknown lvalue 'Delegate' in section 'Service'
Jun 23 10:59:51 client-docker-3 systemd[1]: [/lib/systemd/system/docker.service:24] Unknown lvalue 'Delegate' in section 'Service'
Jun 23 10:59:52 client-docker-3 systemd[1]: [/lib/systemd/system/docker.service:24] Unknown lvalue 'Delegate' in section 'Service' 
```

I'll keep my eyes out for reproducible behaviour, my only best guess is that it has happened after trying to access an image that didn't exist locally and couldn't be pulled from a remote location, but I think that might be coincidence.

**Comment by thaJeztah** [comment]
> `containerd: notify OOM events`

@ccactus Wondering if the machine ran out of memory, causing processes to be killed by the kernel

**Comment by ccactus** [comment]
@thaJeztah Just had it happen again and checked the memory status of the machine, doesn't look like that was an issue. If it's of any help, the issue started re-occuring after we got a "transport is closing" message when building a container and running apt-get.

**Comment by k0nsl** [comment]
Interesting.

I ran into this today while upgrading one of my applications:

```
root@goychat01:~/rocket.chat# bash update-rocketchat.sh
latest: Pulling from rocketchat/rocket.chat
6d827a3ef358: Already exists 
bffd151d9294: Already exists 
d8ab00946d93: Already exists 
4ad00d88ffa0: Already exists 
f7572b3f3d12: Pull complete 
Digest: sha256:b743a6e6ab17f0e2314b00edf8af0313ddee99b68b636527659e403ace6bb632
Status: Downloaded newer image for rocketchat/rocket.chat:latest
Stopping rocketchat_chat_1 ... error

ERROR: for rocketchat_chat_1  Cannot stop container 21b16fbb189fca50bcedbc7efb448935dd0511546aacd0380be0d4c135167f38: Cannot kill container 21b16fbb189fca50bcedbc7efb448935dd0511546aacd0380be0d4c135167f38: rpc error: code = 14 desc = grpc: the connection is unavailable
WARNING: Not including one-off containers created by `docker-compose run`.
To include them, use `docker-compose rm --all`.
This will be the default behavior in the next version of Compose.
```

**Comment by ccactus** [comment]
I'm still seeing this regularly on our machines, in most cases I haven't been able to see anything in the logs that would give an indication of the cause, even with debug turned on. However, the error seems to always be preceded by a "transport is closing" message. When that occurred today the log contained:

Jul 09 03:06:49 client-docker-3 dockerd[6785]: time="2017-07-09T03:06:49.349331797+01:00" level=warning msg="Couldn't run auplink before unmount /var/lib/docker/aufs/mnt/d5fcaeba8b86f031d9f6254dc6241aa63e9aa00d8be36560cde4b95e85a57ef9: signal: seg...fault (core dumped)"
Jul 09 03:06:49 client-docker-3 dockerd[6785]: time="2017-07-09T03:06:49.698734397+01:00" level=warning msg="Couldn't run auplink before unmount /var/lib/docker/aufs/mnt/d5fcaeba8b86f031d9f6254dc6241aa63e9aa00d8be36560cde4b95e85a57ef9: signal: seg...fault (core dumped)"

Unfortunately this is the only one that didn't have debug enabled.

@thaJeztah since this issue has been closed for some time, would you like me to open a new one?

**Comment by thaJeztah** [comment]
@ccactus looks like there's a core dump in there, which may indicate a kernel issue; kernel 3.16 is known to have some issues, so may be at the core of your issue; feel free to open a new issue with details though

**Comment by ccactus** [comment]
Thanks @thaJeztah. If I find any useful logs I might. Any links to mentioned issues for kernel 3.16?

**Comment by thaJeztah** [comment]
I don't have them at hand, but searching for `3.16` and the `area/kernel` tag may provide some. These are issues that are not always easy to reproduce, and only occur "sometimes", unfortunately.

**Comment by datacarl** [comment]
We have this happening frequently on CentOS 7.3 with 17.06.0-ce. Just upgraded the kernel but the problem persists. I don't think we are doing anything out of the ordinary except using namespaces.

This is what happened last time I ran into it on a newly provisioned host.

1) `docker run hello-world`
Outputs the hello-world text but the container doesn't exit and doesn't respond to ctrl-c.

2) Open a new shell and `docker run hello-world`
> docker: Error response from daemon: grpc: the connection is unavailable.

3) `systemctl restart docker` and then `docker run hello-world`
Now everything works again

4) `docker ps -a`

```
f5aed51a6925        hello-world         "/hello"            About a minute ago   Exited (0) About a minute ago                       infallible_poitras
d9f64b59527d        hello-world         "/hello"            10 minutes ago       Created                                             pedantic_varahamihira
a8620847513c        hello-world         "/hello"            12 minutes ago       Exited (0) About a minute ago                       serene_khorana
```

Notice how the first container didn't exit until later and that the container I ran that resulted in "grpc: the connection is unavailable." reached the `Created` state but nothing further.

Daemon log doesn't show anything interesting.

```
Containers: 3
 Running: 0
 Paused: 0
 Stopped: 3
Images: 1
Server Version: 17.06.0-ce
Storage Driver: devicemapper
 Pool Name: docker-thinpool
 Pool Blocksize: 524.3kB
 Base Device Size: 10.74GB
 Backing Filesystem: xfs
 Data file:
 Metadata file:
 Data Space Used: 64.49MB
 Data Space Total: 102GB
 Data Space Available: 101.9GB
 Metadata Space Used: 184.3kB
 Metadata Space Total: 1.07GB
 Metadata Space Available: 1.069GB
 Thin Pool Minimum Free Space: 10.2GB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 0
 Library Version: 1.02.135-RHEL7 (2016-11-16)
Logging Driver: syslog
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: cfb82a876ecc11b5ca0977d1733adbe58599088a
runc version: 2d41c047c83e09a6d61d464906feb2a2f3c52aa4
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
 userns
Kernel Version: 3.10.0-514.26.2.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.518GiB
Name: ip-10-121-151-102
ID: HYSP:F5EW:E6ZO:OAOK:FG5L:YYAB:CSHU:OV3J:OESZ:7IWF:MYBL:YDL6
Docker Root Dir: /var/lib/docker/500000.500000
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 19
 Goroutines: 27
 System Time: 2017-08-04T11:59:55.984632351Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Labels:
 com.<domain>.host=default
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Comment by thaJeztah** [comment]
This issue was closed because we didn't have a way to reproduce the issue in the original reporter's setup.

Unfortunately, the `docker: Error response from daemon: grpc: the connection is unavailable.` error is quite generic. Basically it means: the docker daemon (`dockerd`) was unable to make a gRPC connection with the containerd daemon (`docker-containerd`); in one of the cases above, this was due to a kernel panic, but other situations, such as the system OOM-killing the daemon could have the same result.

Given that there's many possible causes this may happen, I'm locking the conversation on this issue to prevent unrelated issues from being conflated in a single discussion (which complicates triage, and leads to information go missing).

If you're having this issue, and no reasonable explanation why this happens;

- Open a new issue
- Provide the requested information from the issue template
- Make sure to check logs, including system logs, which may contain information about (e.g.) the kernel OOM killing processes (in that case, it's not a bug)

Thanks!
